File: backend/app/extractors/registry.py
"""Registry for running all declaration extractors consistently."""

from __future__ import annotations

from collections.abc import Callable

from app.domain import ExtractedField, ImageMeta, OCRWord, RulesConfig, ScanContext
from app.extractors.best_before import extract_best_before
from app.extractors.common_name import extract_common_name
from app.extractors.consumer_care import extract_consumer_care
from app.extractors.country_origin import extract_country_origin, extract_importer_address
from app.extractors.dimensions import extract_dimensions
from app.extractors.manufacturer import extract_manufacturer_address
from app.extractors.mfg_date import extract_mfg_date
from app.extractors.mrp import extract_mrp
from app.extractors.net_quantity import extract_net_quantity
from app.extractors.unit_price import extract_unit_price

Extractor = Callable[[list[OCRWord], ImageMeta, RulesConfig], ExtractedField | None]
VIRTUAL_FIELDS = frozenset({"listing_declarations", "declaration_readability"})


def _check(rules: RulesConfig, field_name: str):
    return next((check for check in rules.checks if check.field == field_name), None)


def _manufacturer(words: list[OCRWord], meta: ImageMeta, rules: RulesConfig):
    check = _check(rules, "manufacturer_address")
    pin_pattern = check.pin_code_regex if check and check.pin_code_regex else r"\b[1-9][0-9]{5}\b"
    return extract_manufacturer_address(words, meta, pin_pattern)


def _net_quantity(words: list[OCRWord], meta: ImageMeta, rules: RulesConfig):
    check = _check(rules, "net_quantity")
    units = check.requires_unit_in if check and check.requires_unit_in else ["g", "kg", "ml", "l"]
    return extract_net_quantity(words, meta, units)


def _mrp(words: list[OCRWord], meta: ImageMeta, rules: RulesConfig):
    check = _check(rules, "mrp")
    phrase = (
        check.tax_inclusive_phrase_regex
        if check and check.tax_inclusive_phrase_regex
        else r"(?i)\binclusive\s+of\s+all\s+taxes\b"
    )
    return extract_mrp(words, meta, phrase)


def _consumer_care(words: list[OCRWord], meta: ImageMeta, rules: RulesConfig):
    check = _check(rules, "consumer_care")
    email = check.email_regex if check and check.email_regex else r"[^\s@]+@[^\s@]+\.[^\s@]+"
    phone = check.phone_regex if check and check.phone_regex else r"(?:\+91[\s-]?)?[6-9]\d{9}"
    return extract_consumer_care(words, meta, email, phone)


def _manufacture_date(words: list[OCRWord], meta: ImageMeta, rules: RulesConfig):
    check = _check(rules, "mfg_date")
    pattern = check.date_format_regex if check and check.date_format_regex else r"(?i)\bmfg\b.*\d"
    return extract_mfg_date(words, meta, pattern)


def _simple(
    extractor: Callable[[list[OCRWord], ImageMeta], ExtractedField],
) -> Extractor:
    def run(words: list[OCRWord], meta: ImageMeta, rules: RulesConfig) -> ExtractedField:
        return extractor(words, meta)

    return run


EXTRACTORS: dict[str, Extractor] = {
    "manufacturer_address": _manufacturer,
    "net_quantity": _net_quantity,
    "mrp": _mrp,
    "consumer_care": _consumer_care,
    "mfg_date": _manufacture_date,
    "common_name": _simple(extract_common_name),
    "country_origin": _simple(extract_country_origin),
    "best_before": _simple(extract_best_before),
    "dimensions": _simple(extract_dimensions),
    "unit_price": _simple(extract_unit_price),
    "importer_address": _simple(extract_importer_address),
}


def _inside_viewport(word: OCRWord) -> bool:
    x, y, width, height = word.bbox
    return width > 0 and height > 0 and x >= 0 and y >= 0 and x + width <= 1 and y + height <= 1


def extract_all(
    words: list[OCRWord],
    image_meta: ImageMeta,
    context: ScanContext,
    rules: RulesConfig,
) -> dict[str, ExtractedField | None]:
    """Run registered extractors without applying legal applicability policy."""

    visible_words = (
        [word for word in words if _inside_viewport(word)]
        if context.mode == "ecommerce_listing"
        else words
    )
    return {
        field_name: extractor(visible_words, image_meta, rules)
        for field_name, extractor in EXTRACTORS.items()
    }

File: frontend/lib/rules/extractors/registry.ts
import type {
  CheckConfig,
  ExtractedField,
  ImageMeta,
  OCRWord,
  RulesConfig,
  ScanContext,
} from '../domain';
import { extractBestBefore } from './best-before';
import { extractCommonName } from './common-name';
import { extractConsumerCare } from './consumer-care';
import { extractCountryOrigin, extractImporterAddress } from './country-origin';
import { extractDimensions } from './dimensions';
import { extractManufacturerAddress } from './manufacturer';
import { extractMfgDate } from './mfg-date';
import { extractMrp } from './mrp';
import { extractNetQuantity } from './net-quantity';
import { extractUnitPrice } from './unit-price';
import { sortWordsSpatially } from '../../spatial-layout';

export type Extractor = (
  words: OCRWord[],
  meta: ImageMeta,
  rules: RulesConfig
) => ExtractedField | null;

export const VIRTUAL_FIELDS = new Set([
  'listing_declarations',
  'declaration_readability',
]);

function findCheck(rules: RulesConfig, fieldName: string): CheckConfig | undefined {
  return rules.checks.find((c) => c.field === fieldName);
}

function runManufacturer(
  words: OCRWord[],
  meta: ImageMeta,
  rules: RulesConfig
): ExtractedField {
  const check = findCheck(rules, 'manufacturer_address');
  const pinPattern =
    check && check.pin_code_regex ? check.pin_code_regex : '\\b[1-9][0-9]{5}\\b';
  return extractManufacturerAddress(words, meta, pinPattern);
}

function runNetQuantity(
  words: OCRWord[],
  meta: ImageMeta,
  rules: RulesConfig
): ExtractedField {
  const check = findCheck(rules, 'net_quantity');
  const units =
    check && check.requires_unit_in ? check.requires_unit_in : ['g', 'kg', 'ml', 'l'];
  return extractNetQuantity(words, meta, units);
}

function runMrp(
  words: OCRWord[],
  meta: ImageMeta,
  rules: RulesConfig
): ExtractedField {
  const check = findCheck(rules, 'mrp');
  const phrase =
    check && check.tax_inclusive_phrase_regex
      ? check.tax_inclusive_phrase_regex
      : '(?i)\\binclusive\\s+of\\s+all\\s+taxes\\b';
  const extracted = extractMrp(words, meta, phrase);
  if (extracted && extracted.value) {
    return {
      name: extracted.name,
      value: `MRP ${extracted.value} (Inclusive of all taxes)`,
      bbox: extracted.bbox,
      confidence: extracted.confidence,
      evidence_spans: extracted.evidence_spans,
      conflicting_values: extracted.conflicting_values,
    };
  }
  return extracted;
}

function runConsumerCare(
  words: OCRWord[],
  meta: ImageMeta,
  rules: RulesConfig
): ExtractedField {
  const check = findCheck(rules, 'consumer_care');
  const email =
    check && check.email_regex ? check.email_regex : '[^\\s@]+@[^\\s@]+\\.[^\\s@]+';
  const phone =
    check && check.phone_regex ? check.phone_regex : '(?:\\+91[\\s-]?)?[6-9]\\d{9}';
  return extractConsumerCare(words, meta, email, phone);
}

function runMfgDate(
  words: OCRWord[],
  meta: ImageMeta,
  rules: RulesConfig
): ExtractedField {
  const check = findCheck(rules, 'mfg_date');
  const pattern =
    check && check.date_format_regex ? check.date_format_regex : '(?i)\\bmfg\\b.*\\d';
  return extractMfgDate(words, meta, pattern);
}

function wrapSimple(
  extractor: (words: OCRWord[], meta: ImageMeta) => ExtractedField
): Extractor {
  return (words: OCRWord[], meta: ImageMeta) => extractor(words, meta);
}

export const EXTRACTORS: Record<string, Extractor> = {
  manufacturer_address: runManufacturer,
  net_quantity: runNetQuantity,
  mrp: runMrp,
  consumer_care: runConsumerCare,
  mfg_date: runMfgDate,
  common_name: wrapSimple(extractCommonName),
  country_origin: wrapSimple(extractCountryOrigin),
  best_before: wrapSimple(extractBestBefore),
  dimensions: wrapSimple(extractDimensions),
  unit_price: wrapSimple(extractUnitPrice),
  importer_address: wrapSimple(extractImporterAddress),
};

function insideViewport(word: OCRWord): boolean {
  const [x, y, width, height] = word.bbox;
  return (
    width > 0 &&
    height > 0 &&
    x >= 0 &&
    y >= 0 &&
    x + width <= 1 &&
    y + height <= 1
  );
}

export function extractAll(
  words: OCRWord[],
  imageMeta: ImageMeta,
  context: ScanContext,
  rules: RulesConfig
): Record<string, ExtractedField | null> {
  const visibleWords =
    context.mode === 'ecommerce_listing'
      ? words.filter(insideViewport)
      : words;

  // Reorder words into visual row-by-row reading order (top-to-bottom, left-to-right)
  const orderedWords = sortWordsSpatially(visibleWords);
  console.log(`[extractAll] Running ${Object.keys(EXTRACTORS).length} extractors on ${orderedWords.length} spatially ordered tokens...`);

  const result: Record<string, ExtractedField | null> = {};
  for (const [fieldName, extractor] of Object.entries(EXTRACTORS)) {
    const field = extractor(orderedWords, imageMeta, rules);
    result[fieldName] = field;
    if (field && field.value) {
      console.log(`  ✓ [extractAll] ${fieldName}: "${field.value}" (conf: ${(field.confidence * 100).toFixed(0)}%)`);
    } else {
      console.log(`  ✗ [extractAll] ${fieldName}: null`);
    }
  }
  return result;
}
